storage: MySQL Source Purification & Planning

rjobanp opened the pull request:
### Motivation

Part of the work towards a prototype for: https://github.com/MaterializeInc/materialize/issues/15901

Implements basic purification for the MySQL `CREATE SOURCE` statement, with support for a subset of MySQL's scalar types.

This closely follows the purification implementation for the Postgres source, but skips all the table-cast / text-expression logic that may not be necessary with MySQL since we're aiming to ingest binary data during replication. 

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]


    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

* Would recommend reviewing commit-by-commit
* The first commit in this PR is scaffolding added by @petrosagg and is shared with the open PR for the source operator: https://github.com/MaterializeInc/materialize/pull/24203

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

rjobanp commented on src/sql/src/pure.rs:
```diff
@@ -143,6 +146,56 @@ fn subsource_name_gen(
 
+//// Validates the requested subsources do not have name conflicts with each other
+/// and that the same upstream table is not referenced multiple times.
+fn validate_subsource_names<T>(
```
this is pure code-movement from `validate_requested_subsources` in `src/sql/src/pure/postgres.rs` so it can be re-used for mysql.

petrosagg replied:
Amazing! :tada:

mjibson commented on src/sql/src/catalog.rs:
```diff
@@ -1328,6 +1330,52 @@ impl<'a, T> ErsatzCatalog<'a, T> {
+    /// - If the normalized `PartialItemName` does not resolve to an item in
+    ///   `self.0`.
+    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &'a T), PlanError> {
+        let name = normalize::unresolved_item_name(item)?;
```
Should this fn assert anything about `name.database`?

rjobanp replied:
No longer applicable after removing the 2-layer catalog implementation due to comments below

mjibson commented on src/sql/src/pure/mysql.rs:
```diff
@@ -0,0 +1,151 @@
+
+pub(super) fn generate_targeted_subsources<F>(
+    scx: &StatementContext,
+    validated_requested_subsources: Vec<(UnresolvedItemName, UnresolvedItemName, &MySqlTableDesc)>,
```
This should probably be a struct to disambiguate the item names.

mjibson reviewed (approved):
Parser and catalog parts lgtm. Didn't look too hard at purification because it's mostly mysql logic which I think storage folks will own.

benesch reviewed (commented):
SQL parser changes also LGTM. Didn't meaningfully review any of the rest, but on a quick skim it looks fantastic.

petrosagg commented on src/sql/src/catalog.rs:
```diff
@@ -1261,24 +1261,26 @@ impl CatalogError {
 impl Error for CatalogError {}
 
+// Implements a trait with a generic parameter T and a lifetime 'a for that parameter
+pub(crate) trait SubsourceCatalog<'a, T> {
```
You don't need the lifetime here, when something is generic over `T` it means that `T` can become *any* type, including references, which are just another type! This is a common misconception and this is an enlightening read on the topic https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md

I attach the diff that fixes the lifetimes since I tried it out before commenting anyway:

```diff
diff --git a/src/sql/src/catalog.rs b/src/sql/src/catalog.rs
index d099f17e33..3741bad2a1 100644
--- a/src/sql/src/catalog.rs
+++ b/src/sql/src/catalog.rs
@@ -1262,17 +1262,17 @@ impl CatalogError {
 impl Error for CatalogError {}
 
 // Implements a trait with a generic parameter T and a lifetime 'a for that parameter
-pub(crate) trait SubsourceCatalog<'a, T> {
-    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &'a T), PlanError>;
+pub(crate) trait SubsourceCatalog<T> {
+    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &T), PlanError>;
 }
 
 /// Provides a method of generating a 3-layer catalog on the fly, and then
 /// resolving objects within it.
-pub(crate) struct SubsourceCatalogTriple<'a, T>(
-    pub BTreeMap<String, BTreeMap<String, BTreeMap<String, &'a T>>>,
+pub(crate) struct SubsourceCatalogTriple<T>(
+    pub BTreeMap<String, BTreeMap<String, BTreeMap<String, T>>>,
 );
 
-impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogTriple<'a, T> {
+impl<T> SubsourceCatalog<T> for SubsourceCatalogTriple<T> {
     /// Returns the fully qualified name for `item`, as well as the `T` that it
     /// describes.
     ///
@@ -1280,7 +1280,7 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogTriple<'a, T> {
     /// - If `item` cannot be normalized to a [`PartialItemName`]
     /// - If the normalized `PartialItemName` does not resolve to an item in
     ///   `self.0`.
-    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &'a T), PlanError> {
+    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &T), PlanError> {
         let name = normalize::unresolved_item_name(item)?;
 
         let schemas = match self.0.get(&name.item) {
@@ -1314,7 +1314,7 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogTriple<'a, T> {
         };
 
         let desc = match databases.get(database) {
-            Some(desc) => *desc,
+            Some(desc) => desc,
             None => sql_bail!("database {database} not found source"),
         };
 
@@ -1332,9 +1332,9 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogTriple<'a, T> {
 
 /// Provides a method of generating a 2-layer catalog on the fly, and then
 /// resolving objects within it.
-pub(crate) struct SubsourceCatalogDouble<'a, T>(pub BTreeMap<String, BTreeMap<String, &'a T>>);
+pub(crate) struct SubsourceCatalogDouble<T>(pub BTreeMap<String, BTreeMap<String, T>>);
 
-impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogDouble<'a, T> {
+impl<T> SubsourceCatalog<T> for SubsourceCatalogDouble<T> {
     /// Returns the fully qualified name for `item`, as well as the `T` that it
     /// describes.
     ///
@@ -1342,7 +1342,7 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogDouble<'a, T> {
     /// - If `item` cannot be normalized to a [`PartialItemName`]
     /// - If the normalized `PartialItemName` does not resolve to an item in
     ///   `self.0`.
-    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &'a T), PlanError> {
+    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &T), PlanError> {
         let name = normalize::unresolved_item_name(item)?;
 
         let schemas = match self.0.get(&name.item) {
@@ -1361,7 +1361,7 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogDouble<'a, T> {
         };
 
         let desc = match schemas.get(schema) {
-            Some(desc) => *desc,
+            Some(desc) => desc,
             None => sql_bail!("schema {schema} not found in source"),
         };
 
diff --git a/src/sql/src/pure.rs b/src/sql/src/pure.rs
index 61a5d33f71..72e38bc506 100644
--- a/src/sql/src/pure.rs
+++ b/src/sql/src/pure.rs
@@ -80,7 +80,7 @@ pub(crate) mod error;
 mod mysql;
 mod postgres;
 
-fn subsource_gen<'a, T, F: SubsourceCatalog<'a, T>>(
+fn subsource_gen<'a, T, F: SubsourceCatalog<&'a T>>(
     selected_subsources: &mut Vec<CreateSourceSubsource<Aug>>,
     catalog: &F,
     source_name: &UnresolvedItemName,
@@ -117,7 +117,7 @@ fn subsource_gen<'a, T, F: SubsourceCatalog<'a, T>>(
 
         let (qualified_upstream_name, desc) = catalog.resolve(subsource.reference.clone())?;
 
-        validated_requested_subsources.push((qualified_upstream_name, subsource_name, desc));
+        validated_requested_subsources.push((qualified_upstream_name, subsource_name, *desc));
     }
 
     Ok(validated_requested_subsources)
diff --git a/src/sql/src/pure/mysql.rs b/src/sql/src/pure/mysql.rs
index a7382c43d2..8bd72b3d34 100644
--- a/src/sql/src/pure/mysql.rs
+++ b/src/sql/src/pure/mysql.rs
@@ -26,7 +26,7 @@ use crate::plan::{PlanError, StatementContext};
 
 pub(super) fn derive_catalog_from_tables<'a>(
     tables: &'a [MySqlTableDesc],
-) -> Result<SubsourceCatalogDouble<'a, MySqlTableDesc>, PlanError> {
+) -> Result<SubsourceCatalogDouble<&'a MySqlTableDesc>, PlanError> {
     // An index from table name -> schema name -> MySqlTableDesc
     let mut tables_by_name = BTreeMap::new();
     for table in tables.iter() {
diff --git a/src/sql/src/pure/postgres.rs b/src/sql/src/pure/postgres.rs
index 366e3cdadc..fab2b37b35 100644
--- a/src/sql/src/pure/postgres.rs
+++ b/src/sql/src/pure/postgres.rs
@@ -33,7 +33,7 @@ use super::error::PgSourcePurificationError;
 pub(super) fn derive_catalog_from_publication_tables<'a>(
     database: &'a str,
     publication_tables: &'a [PostgresTableDesc],
-) -> Result<SubsourceCatalogTriple<'a, PostgresTableDesc>, PlanError> {
+) -> Result<SubsourceCatalogTriple<&'a PostgresTableDesc>, PlanError> {
     // An index from table name -> schema name -> database name -> PostgresTableDesc
     let mut tables_by_name = BTreeMap::new();
     for table in publication_tables.iter() {
@@ -76,7 +76,7 @@ pub(super) async fn validate_requested_subsources_privileges(
 }
 
 pub(super) fn generate_text_columns(
-    publication_catalog: &SubsourceCatalogTriple<'_, PostgresTableDesc>,
+    publication_catalog: &SubsourceCatalogTriple<&PostgresTableDesc>,
     text_columns: &mut [UnresolvedItemName],
     option_name: &str,
 ) -> Result<BTreeMap<u32, BTreeSet<String>>, PlanError> {

```

rjobanp replied:
That link is very enlightening! Will refactor with your changes to avoid the unnecessary lifetimes

joacoc commented on src/mysql-util/src/schemas.rs:
```diff
@@ -7,52 +7,181 @@
+        conn.exec(table_q, schemas).await?
+    } else {
+        // Get all tables in non-system schemas.
+        // TODO(roshan): Many users create user-defined tables in the `mysql` system schema, since mysql doesn't
```
A Notion task/GH issue for this will help ensure it gets documented. Btw. Is there any documentation task/issue?

rjobanp replied:
@joacoc good point -- I'll plan to add this and other TODOs in these PRs as tasks in our notion doc today https://www.notion.so/materialize/MySQL-Source-Prototype-8838b49eb2124e3580a4e9922761d423

>  Btw. Is there any documentation task/issue?
There is not yet! I can create that too, though this is currently a 'prototype' feature so there is no urgency at the moment for comprehensive docs.

petrosagg commented on src/sql/src/catalog.rs:
```diff
@@ -1328,6 +1330,52 @@ impl<'a, T> ErsatzCatalog<'a, T> {
 
+/// Provides a method of generating a 2-layer catalog on the fly, and then
+/// resolving objects within it.
+pub(crate) struct SubsourceCatalogDouble<'a, T>(pub BTreeMap<String, BTreeMap<String, &'a T>>);
```
I think this double version goes a bit against the abstraction (that granted is not well spelled out anywhere yet). The idea is that subsource planning should be handled identically by requiring that each source implementation defines and exports a subsource catalog with a canonical structure. This structure must be in the form of a three layered catalog even if the upstream source has a simpler structure. In that case we can generate a default name for the layer that's missing.

As an example the load generator source, for which no upstream system exists, but it still reports a fully fledged three layer catalog. Can something like this be done for MySQL?

petrosagg replied:
After doing that you can also only keep a `SubsourceCatalog<T>` struct and get rid of the trait

rjobanp replied:
> I would go for a default name like "mysql" or something like that. 

Hmm, then that would leak `mysql` details into the `impl SubsourceCatalog` implementation of `resolve`..

> The resolve method should already support resolving names based on just the item name or just the item and the schema so long as they are non-ambiguous

Yep it supports that, but we don't want to return a new `UnresolvedItemName` with the extra `mysql` prefix in there, so I think we still need to check if the input only contained 2 `Ident`s and used a default catalog layer, right?

rjobanp replied:
Okay got it - thanks for clarifying. We'll have to be careful to resolve things the same way on the source side, so I'll tackle that when I rebase #24203 against this PR. I've implemented this in the commit I'm about to push.

petrosagg commented on src/sql/src/pure/mysql.rs:
```diff
@@ -0,0 +1,151 @@
+
+            // TODO: The 'column_key' value returned by MySQL's information_schema.columns table only
+            // tells us if a column is the first-column in a multi-column key. We should also check
+            // the information_schema.statistics table to find secondary columns in a multi-column index.
```
If we do not have the full picture of the primary key then we should not propagate it to our schema because that will lead to miscompilations of queries since the optimizer  will operate under false assumptions

petrosagg commented on src/storage-types/src/sources/mysql.rs:
```diff
@@ -0,0 +1,163 @@
+pub static MYSQL_PROGRESS_DESC: Lazy<RelationDesc> = Lazy::new(|| {
+    RelationDesc::empty()
+        // TODO(petrosagg): UUIDs in ranges are not supported and additionally the
+        // Partitioned timestamp type forces you to deal with open/closed ranges which are
```
As of https://github.com/MaterializeInc/materialize/pull/24233 this is now fixed so we could set this to be:

```
RelationDesc::empty()
    .with_column("source_id_lower", ScalarType::Uuid.nullable(false))
    .with_column("source_id_upper", ScalarType::Uuid.nullable(false))
    .with_column("transaction_id", ScalarType::UInt64.nullable(true)) // <-- notice this changed too
```

petrosagg commented on src/workspace-hack/Cargo.toml:
```diff
@@ -54,7 +54,8 @@ futures-util = { version = "0.3.28", features = ["channel", "io", "sink"] }
 hyper = { version = "0.14.27", features = ["full"] }
-indexmap = { version = "1.9.1", default-features = false, features = ["std"] }
+indexmap-dff4ba8e3ae991db = { package = "indexmap", version = "1.9.1", default-features = false, features = ["std"] }
+indexmap-f595c2ba2a3f28df = { package = "indexmap", version = "2.0.0" }
```
Something is off here

petrosagg reviewed (commented):
Really exciting! Left quite a bit of comments but all of them are minor, great work!

nrainer-materialize commented:
@rjobanp: Please let QA know if there is anything that can and should already be tested. 😀

rjobanp commented on src/mysql-util/src/schemas.rs:
```diff
@@ -7,52 +7,181 @@
-                "char" => MySqlDataType::Varchar(usize::MAX),
-                _ => return Err(MySqlError::UnsupportedDataType(data_type)),
+        for mut row in column_rows {
+            let name = row.take(INFO_SCHEMA_COLS["column_name"]).unwrap();
```
Ah yes, I saw that too and originally implemented all of this using `row.take("column_name").unwrap()`, etc but unfortunately hit runtime errors -- `take()` would always return `None` when using the `&str` index-type. Based on your other comment I plan to switch to specifying the types directly -- I thought this might be more ergonomic but it doesn't look that way anymore.

Separately I do want to figure out why the `Row.take(&str)` method isn't working correctly -- I may file an issue with `mysql-async`

rjobanp commented on src/mysql-util/src/desc.rs:
```diff
@@ -21,20 +29,139 @@ pub struct MySqlTableDesc {
+/// https://dev.mysql.com/doc/refman/8.0/en/information-schema-columns-table.html
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub enum MySqlColumnKey {
+    PRI,
```
After more investigation into mysql constraints I'm going to remove this `column_key` field as it's pretty useless since it doesn't tell provide enough information about multi-column indexes.

Instead I've figured out how to properly query the `information_schema.statistics` table to get the appropriate unique index info, and will represent those as `keys` on the `MySqlTableDesc` similar to how we do for Postgres

rjobanp commented on src/mysql-util/src/schemas.rs:
```diff
@@ -7,52 +7,181 @@
+                    precision: datetime_precision.and_then(|f| f.try_into().ok()),
+                },
+                "time" => ScalarType::Time,
+                "year" => ScalarType::Int32,
```
sounds good, I will remove `year` for now. I agree that without explicit syntax supporting this will likely be confusing for users

rjobanp commented on src/mysql-util/src/schemas.rs:
```diff
@@ -7,52 +7,181 @@
+                "time" => ScalarType::Time,
+                "year" => ScalarType::Int32,
+                "decimal" => ScalarType::Numeric {
+                    max_scale: numeric_scale.and_then(|f| f.try_into().ok()),
```
Based on:
https://github.com/MaterializeInc/materialize/blob/6b05a170381bc2d6ac325043cde33b25f7baecd5/src/repr/src/scalar.rs#L1382-L1393

our max-precision is always `NUMERIC_DATUM_MAX_PRECISION` and we only check the max-scale if it's not `None`. So if the MySQL type is `NUMERIC(5, 2)` we can represent it as `ScalarType::Numeric { max_scale: 2 }` as long as `5 < NUMERIC_DATUM_MAX_PRECISION`.
I'm also not sure if we're always storing `NUMERIC_DATUM_MAX_PRECISION` number of digits to represent all `Numeric` values? Or if we can store less digits when that much precision isn't necessary?

rjobanp commented on src/mysql-util/src/schemas.rs:
```diff
@@ -7,52 +7,181 @@
+                "float" => ScalarType::Float32,
+                "double" => ScalarType::Float64,
+                "timestamp" => ScalarType::Timestamp {
+                    precision: datetime_precision.and_then(|f| f.try_into().ok()),
```
good point, will make fallible

rjobanp commented on src/mysql-util/src/schemas.rs:
```diff
@@ -7,52 +7,181 @@
+                "time" => ScalarType::Time,
+                "year" => ScalarType::Int32,
+                "decimal" => ScalarType::Numeric {
+                    max_scale: numeric_scale.and_then(|f| f.try_into().ok()),
```
It's checked in the `NumericMaxScale` `TryFrom` implementation (which I'll stop ignoring the error for). But I also realized that I'm not retrieving or checking the `numeric_precision` returned for this column at all, so I will also add a check to validate that.

rjobanp commented:
Thanks for the reviews everyone; I think I've addressed all the PR feedback thus far (in the last 4 commits)

@nrainer-materialize there are basic tests here for validating purification and sub-source naming (similar to postgres). Once this and the other PR https://github.com/MaterializeInc/materialize/pull/24203 are merged then I think we can start working on more comprehensive tests to validate replication, mysql type decoding, etc.

rjobanp commented on src/sql-parser/src/ast/defs/ddl.rs:
```diff
@@ -911,7 +911,8 @@ impl_display_t!(KafkaSinkConfigOption);
 pub enum PgConfigOptionName {
-    /// Hex encoded string of binary serialization of `dataflow_types::PostgresSourceDetails`
+    /// Hex encoded string of binary serialization of
+    /// [`mz_storage_types::sources::postgres::PostgresSourcePublicationDetails`]
```
@petrosagg this is erroring on
```
error: unresolved link to `mz_storage_types::sources::postgres::PostgresSourcePublicationDetails`
--
  | 2024-01-10 20:21:13 UTC | --> src/sql-parser/src/ast/defs/ddl.rs:915:11
  | 2024-01-10 20:21:13 UTC | \|
  | 2024-01-10 20:21:13 UTC | 915 \|     /// [`mz_storage_types::sources::postgres::PostgresSourcePublicationDetails`]
  | 2024-01-10 20:21:13 UTC | \|           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item named `mz_storage_types` in scope
  | 2024-01-10 20:21:13 UTC | \|
```

I'm a little hesitant to add `mz-storage-types` as a dependency of `mz-sql-parser` to satisfy this -- how do you feel about me removing the doc link format here (and for the mysql one below)?

petrosagg replied:
Sounds fine, remove the formatting :+1:

nrainer-materialize commented:
> @nrainer-materialize there are basic tests here for validating purification and sub-source naming (similar to postgres). Once this and the other PR #24203 are merged then I think we can start working on more comprehensive tests to validate replication, mysql type decoding, etc.

Sounds good 👍

rjobanp commented on src/mysql-util/src/schemas.rs:
```diff
@@ -7,57 +7,241 @@
+    Option<i64>, // numeric_scale
+    Option<i64>, // datetime_precision
+    Option<i64>, // character_maximum_length
+);
```
Awesome thanks for the diff -- looks much cleaner!

rjobanp commented on src/mysql-util/src/schemas.rs:
```diff
@@ -7,57 +7,241 @@
+                SELECT
+                    index_name,
+                    group_concat(column_name ORDER BY seq_in_index separator ', ') AS columns,
+                    group_concat(expression ORDER BY seq_in_index separator ', ') AS expressions
```
Fair point -- how do you feel about using `JSON_ARRAYAGG` instead? https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions.html#function_json-arrayagg

It seems easier to do the grouping and filtering (of indexes with expressions) in the database instead, and `JSON_ARRAYAGG` seems to avoid the potential escaping / string concatenation issues that we might otherwise hit.

rjobanp replied:
discussed offline -- Petros suggested a better approach:

```
SELECT
    index_name,
    column_name
FROM information_schema.statistics outer
WHERE
    table_schema NOT IN ('information_schema', 'mysql', 'performance_schema', 'sys')
    AND NOT EXISTS (
        SELECT 1
        FROM information_schema.statistics inner
        WHERE outer.index_name = inner.index_name AND inner.column_name IS NULL
    )
    AND non_unique = 0
    AND table_name = ?
    AND table_schema = ?
ORDER BY index_name, seq_in_index
```
